### PR TITLE
Moving to TS

### PR DESCRIPTION
**Pull request by Ofigelov**
a lot of changes
-all code  was moved in typescript notation
-was removed useless methods like `isGe`, `isEqual` and etc
-was updated `is` method via enums and keyof
-init method was removed, we'll use just constructor for initialization
-removed defensive code, because we have ts
-keeping and sorting viewports were simplified
-demo and standalone versions were removed, so will just one transpiling version like in dcFactory

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -53,28 +53,38 @@ <h2>Simple example</h2>
 
             <textarea style="width: 100%" id="logs" rows="5"></textarea>
 
-            <script data-live-highlight="simple">
+            <script data-live-highlight="simple" defer>
```
Comment (seregap) [inline index.html]: @Ofigelov `defer` won't work with inline scripts, why did you add it?

File: babel.config.js [hunk 2]
```diff
@@ -3,18 +3,15 @@ module.exports = {
         [
             '@babel/preset-env',
             {
-                useBuiltIns: 'usage',
```
Comment (seregap) [inline babel.config.js:6]: @Ofigelov I think we still need `useBuiltIns: 'entry'` to be added

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    private _debounceTimeoutId: number = 0;
+    private _currentDevice: DeviceInfo<Devices> = { size: 0, name: '', index: 0 };
+    private _options: ObserverOptions = { resizeDebounce: DEFAULT_RESIZE_DEBOUNCE, mobileFirst: true };
+
+    constructor(devices: Devices, options?: ObserverOptions) {
+        this._devices = this._sortDevices(devices);
+        if (options) this._options = Object.assign(this._options, options);
+        const device = this._getCurrentDevice;
+        if (device) this._currentDevice = device;
+        window.addEventListener('resize', this._onWindowResize);
```
Comment (seregap) [inline src/index.ts:49]: @Ofigelov as a performance improvement can we look into changing `resize` listener with `matchMedia.onchange`, see [this approach](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Testing_media_queries#Receiving_query_notifications) and [this](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/onchange)
this way we can remove debounce and do perform checks only against real media query change

File: package.json [hunk 4]
```diff
@@ -13,21 +13,37 @@
...
-    "webpack-cli": "^3.3.6"
-  }
+    "@babel/core": "7.10.3",
+    "@babel/plugin-proposal-class-properties": "7.10.1",
+    "@babel/plugin-proposal-object-rest-spread": "7.10.3",
+    "@babel/plugin-syntax-dynamic-import": "7.0.0",
+    "@babel/plugin-transform-object-assign": "7.10.3",
+    "@babel/preset-env": "7.10.3",
+    "@babel/preset-typescript": "7.12.1",
+    "@babel/runtime-corejs3": "7.12.1",
```
Comment (seregap) [inline package.json]: @Ofigelov do you need this dependency?

**Comment by seregap** [comment]
@Ofigelov README.md needs to be changed as well, there are `isGe` etc methods which were removed

File: src/index.ts (hunk 3 above)
Comment (Ofigelov) [reply src/index.ts:49]: `change` event of matchMedia doesn't available in IE11
I think that having the ability to subscribe singlepointly on viewport resize can be helpfull in different situations, for example, to keep the variable `--viewport-height` with actual value.

File: index.html (hunk 1 above)
Comment (Ofigelov) [reply index.html]: My bad, I'll remove it

File: package.json (hunk 4 above)
Comment (Ofigelov) [reply package.json]: Looks like no. I'll remove it.

File: babel.config.js (hunk 2 above)
Comment (Ofigelov) [reply babel.config.js:6]: done

**Review by bububo — APPROVED** [review]

File: package.json [hunk 5]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "@deleteagency/device-observer",
-  "version": "2.0.0",
+  "version": "3.0.0",
   "description": "",
-  "main": "src/index.js",
+  "main": "src/index.ts",
```
Comment (bububo) [inline package.json]: should be path to dist file

File: package.json [hunk 6]
```diff
@@ -1,8 +1,8 @@
...
   "files": [
```
Comment (bububo) [inline package.json]: please remove "files" property or add all files for npm

File: webpack.base.config.js [hunk 7]
```diff
@@ -1,33 +1,45 @@
...
     return {
-        mode: "production",
+        mode: 'production',
         entry: {
-            "device-observer": ["./src/index.js"]
+            'index': './src/index.ts'
         },
         output: {
-            filename: "[name].min.js",
+            filename: './[name].min.js',
```
Comment (bububo) [inline webpack.base.config.js]: the *.d.ts. file should have the same name as result js file (min postfix should be removed)

**Review by seregap — APPROVED** [review]
